# PROYECTO MÓDULO 1

**Integrantes:**
- José Eddie Aguilar Ceballos
- Nancy Esmeralda González Monterrubio
- Ximena Arroyo Ruvalcaba

## Brokers Latinos


**Objetivo General:**

El objetivo de este proyecto es mostrar un programa que puede ayudar a algún inversionista a maximizar la ganancia anual esperada, y de esta manera de sus ganancias, dependiendo de las empresas en las que éste quiere invertir.

Para esto se utilizará programación lineal, a fin de optimizar.

**Objetivo Específico:**

Hay un inversionista en México que quiere invertir en la Bolsa Mexicana de Valores en siete diferentes empresas.

Cuenta con un capital máximo de $500,000.00 (quinientos mil pesos 00/100 M.N.) y quiere dividir su inversión en las siguientes empresas:

- Alsea, S.A.B. de C.V. 
- Megacable Holdings, S.A.B de C.V.
- América Móvil, S.A.B de C.V. 
- Grupo Bimbo, S.A.B de C.V. 
- Arca Continental, S.A.B de C.V. 
- Grupo Financiero Banorte, S.A.B de C.V. 
- Grupo Carso, S.A.B de C.V. 

El inversionista desea tener un número de acciones definidas en las siguientes empresas:

- Arca Continental, S.A.B de C.V. (300 acciones)
- Grupo Financiero Banorte, S.A.B de C.V. (350 acciones)
- Grupo Carso, S.A.B de C.V. (450 acciones)

Además de eso, quiere tener un mínimo de acciones en las siguientes empresas:

- Megacable Holdings, S.A.B de C.V. (900 acciones)
- Alsea, S.A.B. de C.V. (700 acciones)

Y un máximo de acciones en la siguiente empresa:

- América Móvil, S.A.B de C.V. (1400 acciones)

Por último, desea tener un mínimo de acciones en diferentes sectores del mercado bursátil:

- Servicio de telecomunicaciones (2500 acciones)
- Productos de consumo frecuente (1000 acciones)

El objetivo será maximizar el valor de su portafolio en base a lo que quiere invertir.

**El precio de las acciones está representado en la siguiente tabla, la cual fue recuperado de Grupo BMV:**

Empresa|Precio de acción|Sector
:----|----|----
Alsea|$\$$38.46|Servicios y bienes de consumo no básico
Megacable|$\$$45.14|Servicio de Telecomunicaciones
América móvil|$\$$17.03|Servicio de Telecomunicaciones
Grupo Bimbo|$\$$73.45|Productos de consumo frecuente
Arca continental|$\$$139.96|Productos de consumo frecuente
Grupo Financiero Banorte|$\$$133.94|Servicios Financieros
Grupo Carso|$\$$75.19|Industrial

Dentro del mercado bursátil, se utilizan herramientas llamadas múltiplos bursátiles, los cuales se basan en las tendencias de las acciones para brindar información útil acerca de los posibles sucesos de las acciones en el futuro. Esto debido a que no hay una forma exacta de saber la ganancia esperada de una acción.

Una de estas herramientas es el _precio utilidad por acción (P/U)_ el cual se obtiene de dividir el market cap (valor de la empresa) entre la utilidad anual de la empresa. Esta herramienta señala los años aproximados en que tardarías en recuperar el 100% de tu inversión.

$$P/U = \frac{Market cap}{Utilidades}$$

A su vez, el recíproco de este representa, en vía utilidades, cuánto se generó en rendimiento. Por lo tanto, se puede tomar en cuenta esta tasa de rendimiento para predecir la ganancia que se espera de cada acción

$$\text{Tasa de rendimiento} = \frac{1}{P/U}$$

Por lo tanto, en la siguiente tabla se muestran las tasas de rendimiento calculadas de cada acción, el market cap fue recuperado de yahoo finance:

Empresa|Market Cap|Utilidad|Precio utilidad por acción|Tasa de rendimiento
:----|----|----|----|----
Alsea|$\$$31,634,250,887.04|$\$$736,864,000.00|42.930922|2.3229%
Megacable|$\$$113,400,000,000.00|$\$$2,186,481,000.00|51.864160|1.9281%
América móvil|$\$$725,251,528,520.48|$\$$44,480,540,000.00|16.304917|6.1331%	
Grupo Bimbo|$\$$319,210,000,000.00|$\$$10,610,933,000.00|30.083123|3.3241%	
Arca continental|$\$$243,100,000,000.00|$\$$7,316,255,000.00|33.227382|3.0096%
Grupo Financiero Banorte|$\$$386,210,176,200.36|$\$$22,271,777,634.00|17.340788|5.7668%
Grupo Carso|$\$$168,510,000,000.00|$\$$6,464,580,000.00|26.066659|3.8363%	

### Modelo que representa el problema

**1. Criterio de optimización:**

- Maximizar

**2. Función:** 

La función a optimizar es el monto invertido en acciones de cada empresa multiplicado por la tasa de rendimiento anual. Se desea maximizar dicha función para llegar a la máxima ganancia posible.

$$max. 0.023x_1 + 0.019x_2 + 0.061x_3 + 0.033x_4 + 0.030x_5 + 0.057x_6 + 0.038x_7$$


**3. Definir Variables:**

Cada x representa el monto invertido en acciones de las siguientes empresas:

- $x_1$: Alsea, S.A.B. de C.V.
- $x_2$: Megacable Holdings, S.A.B de C.V.
- $x_3$: América Móvil, S.A.B de C.V.
- $x_4$: Grupo Bimbo, S.A.B de C.V.
- $x_5$: Arca Continental, S.A.B de C.V.
- $x_6$: Grupo Financiero Banorte, S.A.B de C.V.
- $x_7$: Grupo Carso, S.A.B de C.V.

**4. Restricciones:**

El monto total de todas las acciones no debe superar los \$500,000
- $x_{1} + x_{2} + x_{3} + x_{4} + x_{5} + x_{6} + x_{7}\leq 500000$

El número de acciones (monto entre precio de acción) de ciertas empresas (Arca continental, Grupo Financiero Banorte y Grupo Carso) deben ser un valor único, por lo tanto:
- $\frac{x_{5}}{139.96} = 300$
- $\frac{x_{6}}{133.94} = 350$
- $\frac{x_{7}}{75.19} = 450$

El número de acciones de ciertas empresas (Alsea y Megacable) debe ser como mínimo: 
- $\frac{x_{2}}{45.14} \geq 900$
- $\frac{x_{1}}{38.46} \geq 700$

El número de acciones de América móvil no debe de superar un número en específico:
- $\frac{x_{3}}{17.03} \leq 1400$

El número de acciones en el sector de Telecomunicaciones (Megacable y América móvil) debe ser como mínimo 2500
- $(\frac{x_{2}}{45.14}+\frac{x_{3}}{17.03}) \geq 2500$

El número de acciones en el sector de Consumos frecuentes (Grupo Bimbo y Arca continental) debe ser como mínimo 1000
- $(\frac{x_{4}}{73.45}+\frac{x_{5}}{139.96}) \geq 1000$




5. Matrices
\begin{gather}
x
 =
  \begin{bmatrix}
   x_1 \\
   x_2 \\
   x_3 \\
   x_4 \\
   x_5 \\
   x_6 \\
   x_7 \\
   \end{bmatrix}
\end{gather}

\begin{gather}
c
 =
  \begin{bmatrix}
   0.023 \\
   0.019 \\
   0.061 \\
   0.033 \\
   0.030 \\
   0.057 \\
   0.038 \\
   \end{bmatrix}
\end{gather}

\begin{gather}
A_{ub}
 =
  \begin{bmatrix}
   1 & 1 & 1 & 1 & 1 & 1 & 1\\
   0 & \frac{-1}{45.14} & 0 & 0 & 0 & 0 & 0\\
   \frac{-1}{38.46} & 0 & 0 & 0 & 0 & 0 & 0\\
   0 & 0 & \frac{1}{17.03} & 0 & 0 & 0 & 0\\
   0 & \frac{-1}{45.14} & \frac{-1}{17.03} & 0 & 0 & 0 & 0\\
   0 & 0 & 0 & \frac{-1}{73.45} & \frac{-1}{139.96} & 0 & 0\\
   \end{bmatrix}
\end{gather}

\begin{gather}
b_{ub}
 =
  \begin{bmatrix}
   500000\\
   -900\\
   -700\\
   1400\\
   -2500\\
   -1000\\
   \end{bmatrix}
\end{gather}

\begin{gather}
A_{eq}
 =
  \begin{bmatrix}
   0 & 0 & 0 & 0 & \frac{1}{139.96} & 0 & 0\\
   0 & 0 & 0 & 0 & 0 & \frac{1}{133.94} & 0\\
   0 & 0 & 0 & 0 & 0 & 0 & \frac{1}{75.19}\
   \end{bmatrix}
\end{gather}

\begin{gather}
b_{eq}
 =
  \begin{bmatrix}
   300\\
   350\\
   450\\
   \end{bmatrix}
\end{gather}


### Solución del problema de optimización

In [2]:
import numpy as np
import scipy.optimize as opt

In [3]:
empresas = ["Alsea", "Megacable", "America móvil", "Grupo Bimbo", "Arca continental", "Grupo Financiero Banorte", "Grupo Carso"]
c = np.array([-0.023, -0.019, -0.061, -0.033, -0.03, -0.057, -0.038])
A_ub = np.array([[1, 1, 1, 1, 1, 1, 1],
                [0, -1/45.14, 0, 0, 0, 0, 0],
                [-1/38.46, 0, 0, 0, 0, 0, 0],
                [0, 0, 1/17.03, 0, 0, 0, 0],
                [0, -1/45.14, -1/17.03, 0, 0, 0, 0],
                [0, 0, 0, -1/73.45, -1/139.96, 0, 0]])
b_ub = np.array([500000, -900, -700, 1400, -2500, -1000])
A_eq = np.array([[0, 0, 0, 0, 1/139.96, 0, 0],
                [0, 0, 0, 0, 0, 1/133.94, 0],
                [0, 0, 1, 0, 0, 0, 1/75.19]])
b_eq = np.array([300, 350, 450])
sol = opt.linprog(c, A_ub, b_ub, A_eq, b_eq)
print("Ganancia anual esperada: $ {}".format(round(sol.get("fun"), 2)*-1))
print("")
for i in range(7):
    print("${:7} invertido en {}".format(round(sol.get("x")[i]), empresas[i]))

Ganancia anual esperada: $ 15819.19

$  26922 invertido en Alsea
$ 112850 invertido en Megacable
$      0 invertido en America móvil
$ 237525 invertido en Grupo Bimbo
$  41988 invertido en Arca continental
$  46879 invertido en Grupo Financiero Banorte
$  33835 invertido en Grupo Carso


Por lo tanto, el número de acciones a invertir en cada empresa son los siguientes:

- Alsea: 700 acciones
- Megacable: 2500 acciones
- América móvil: 0 acciones
- Grupo Bimbo: 3233 acciones
- Arca continental: 300 acciones
- Grupo Financiero Banorte: 350 acciones
- Grupo Carso: 450 acciones

El algoritmo de optimización se condujo a una solución factible debido a que, los resultados obtenidos representan el monto a invertir en cada empresa respetando las cláusulas presentadas por el inversionista, y al mismo tiempo, maximizando su ganancia anual esperada.

Empresa|Esperado (acciones)|Obtenido (acciones)
:----|----|----
Alsea|Mínimo 700 |700
Megacable|Mínimo 900|2500
América móvil|Máximo 1400|0
Grupo Bimbo| - |3233
Arca continental|300|300 
Grupo Financiero Banorte|350|350
Grupo Carso|450|450

Sector|Esperado (acciones)|Obtenido (acciones)
:----|----|----
Servicios de Telecomunicaciones|Mínimo 2500 |2500
Productos de consumo frecuente|Mínimo 1000|3533

**Resultados:**

Empresa|Acciones|Precio|Monto|Tasa rendimiento|Ganancia anual
:----|----|----|----|----|----
Alsea|700|$\$$38.46|$\$$26,922|2.32%|$\$$619.20
Megacable|2500|$\$$45.14|$\$$112,850|1.92%|$\$$2,166.72
América móvil|0|$\$$17.03|$\$$0|6.13%|$\$$0
Grupo Bimbo|3233|$\$$73.45|$\$$237,525|3.32%|$\$$7,838.32
Arca continental|300|$\$$139.96|$\$$41,988|3.00%|$\$$1,256.64
Grupo Financiero Banorte|350|$\$$133.94|$\$$46,879|5.76%|$\$$2,700.23
Grupo Carso|450|$\$$75.19|$\$$33,835|3.83%|$\$$1,295.88
|||||$\$$15,819.19

Como conclusión, se logró maximizar la ganancia del inversionista, esto con base en las restricciones que él mismo propuso en un primer momento. Tanto las empresas en las que obligatoriamente quería invertir, como el mínimo y máximo de acciones en distintos sectores. De esta forma, se obtiene un portafolio de inversión diverso, con diferentes empresas y la máxima ganancia posible.

Sin embargo, consideramos que a pesar de que se maximizó la ganancia esperada con las cláusulas dadas por el cliente, no hay necesidad de invertir un número tan específico de acciones en las empresas que él deseaba, de esta forma, hay otras maneras de invertir para conseguir un portafolio diverso y tal vez con más ganancias.

También hay que tener en cuenta que los múltiplos bursátiles, en este caso el usado Precio Utilidad, están basados en datos e información del pasado, por lo que no pueden ser tan acertados para predecir el futuro de las acciones y sus valores. Debido a esto, la tasa de rendimiento usada es calculada a partir de múltiplos bursátiles, por lo cual su uso dentro del ejercicio es una predicción de la ganancia, la cual puede variar demasiado durante el año de vida de la acción. Además en el mundo bursátil, las acciones siempre cambian debido al comportamiento humano, por lo tanto pueden llegar a ser impredecibles.

**Referencias:**

- Grupo BMV Home. (s. f.). Recuperado 23 de septiembre de 2022, de https://www.bmv.com.mx/
- Yahoo Finance. (s. f.). Recuperado 23 de septiembre de 2022, de https://finance.yahoo.com/